src/query/storages/orc/src/source.rs

```diff
@@ -0,0 +1,112 @@
+// Copyright 2021 Datafuse Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::sync::Arc;
+
+use databend_common_catalog::query_kind::QueryKind;
+use databend_common_catalog::table_context::TableContext;
+use databend_common_exception::ErrorCode;
+use databend_common_exception::Result;
+use databend_common_expression::DataBlock;
+use databend_common_expression::DataSchema;
+use databend_common_pipeline_core::processors::OutputPort;
+use databend_common_pipeline_core::processors::ProcessorPtr;
+use databend_common_pipeline_sources::SyncSource;
+use databend_common_pipeline_sources::SyncSourcer;
+use databend_common_storage::CopyStatus;
+use databend_common_storage::FileStatus;
+use opendal::layers::BlockingLayer;
+use opendal::Operator;
+use orc_rust::ArrowReader;
+use orc_rust::ArrowReaderBuilder;
+
+use super::file_reader::OrcFileReader;
+use crate::one_file_partition::OrcFilePartition;
+
+pub struct ORCSource {
+    table_ctx: Arc<dyn TableContext>,
+    op: Operator,
+    reader: Option<(String, ArrowReader<OrcFileReader>)>,
+    data_schema: Arc<DataSchema>,
+
+    copy_status: Option<Arc<CopyStatus>>,
+}
+
+impl ORCSource {
+    pub fn try_create(
+        output: Arc<OutputPort>,
+        table_ctx: Arc<dyn TableContext>,
+        op: Operator,
+        data_schema: Arc<DataSchema>,
+    ) -> Result<ProcessorPtr> {
+        let copy_status = if matches!(table_ctx.get_query_kind(), QueryKind::CopyIntoTable) {
+            Some(table_ctx.get_copy_status())
+        } else {
+            None
+        };
+        SyncSourcer::create(table_ctx.clone(), output, ORCSource {
+            table_ctx,
+            op,
+            data_schema,
+            reader: None,
+            copy_status,
+        })
+    }
+}
+
+impl SyncSource for ORCSource {
+    const NAME: &'static str = "ORCSource";
+
+    fn generate(&mut self) -> Result<Option<DataBlock>> {
+        if self.reader.is_none() {
+            let part = match self.table_ctx.get_partition() {
+                Some(part) => part,
+                None => return Ok(None),
+            };
+            let file = OrcFilePartition::from_part(&part)?.clone();
+            let path = file.path.clone();
+
+            let file = OrcFileReader {
+                operator: self.op.clone().layer(BlockingLayer::create()?).blocking(),
```

Comment:
Please note this `create` call must happen in async content.

If we do want to call this inside blocking function, please make sure we are guarded by `handle.enter()`, for example:

```rust
use once_cell::sync::Lazy;
use opendal::layers::BlockingLayer;
use opendal::services::S3;
use opendal::BlockingOperator;
use opendal::Operator;
use opendal::Result;

static RUNTIME: Lazy<tokio::runtime::Runtime> = Lazy::new(|| {
    tokio::runtime::Builder::new_multi_thread()
        .enable_all()
        .build()
        .unwrap()
});
///

fn main() -> Result<()> {
    // Create fs backend builder.
    let mut builder = S3::default();
    builder.bucket("test");
    builder.region("us-east-1");

    // Fetch the `EnterGuard` from global runtime.
    let _guard = RUNTIME.enter();
    // Build an `BlockingOperator` with blocking layer to start operating the storage.
    let _: BlockingOperator = Operator::new(builder)?
        .layer(BlockingLayer::create()?)
        .finish()
        .blocking();

    Ok(())
}
```

We can use `GlobalRuntime`.

Reply:
@Xuanwo the one in async panic:

```
     "Cannot start a runtime from within a runtime. This happens \
            because a function (like `block_on`) attempted to block the \
            current thread while the thread is being used to drive \
            asynchronous tasks."
```